File: Design-Patterns-Using-Swift/Flyweight/Game.swift
import Foundation

// Flyweight protocol
protocol Flyweight {
    var intrinsicState: String { get }
    func operation(extrinsicState: String)
}

// Concrete Flyweight class
class ConcreteFlyweight: Flyweight {
    let intrinsicState: String

    init(intrinsicState: String) {
        self.intrinsicState = intrinsicState
    }

    func operation(extrinsicState: String) {
        print("Operation inside concrete flyweight.")
    }
}

// FlyweightFactory class
class FlyweightFactory {
    private var flyweights: [String: Flyweight] = [:]

    func getFlyweight(intrinsicState: String) -> Flyweight {
        if let existingFlyweight = flyweights[intrinsicState] {
            return existingFlyweight
        } else {
            let newFlyweight = ConcreteFlyweight(intrinsicState: intrinsicState)
            flyweights[intrinsicState] = newFlyweight
            return newFlyweight
        }
    }
}

// ClientClass
class ClientClass {
    var flyweight: Flyweight
    var extrinsicState: String

    init(factory: FlyweightFactory, intrinsicState: String, extrinsicState: String) {
        self.flyweight = factory.getFlyweight(intrinsicState: intrinsicState)
        self.extrinsicState = extrinsicState
    }

    func operation() {
        print("Operation inside context.")
        flyweight.operation(extrinsicState: extrinsicState)
    }
}

// Client code
let factory = FlyweightFactory()
let c = ClientClass(factory: factory, intrinsicState: "common", extrinsicState: "separate1")
c.operation()

let c2 = ClientClass(factory: factory, intrinsicState: "common", extrinsicState: "separate2")
c2.operation()

print(c, c2)
print(c.flyweight, c2.flyweight)

/*
Operation inside context.
Operation inside concrete flyweight.
Operation inside context.
Operation inside concrete flyweight.
Game.ClientClass Game.ClientClass
Game.ConcreteFlyweight Game.ConcreteFlyweight
*/

File: Design-Patterns-Using-Ruby/Flyweight/Game.rb
# Flyweight interface
class Flyweight
    attr_reader :intrinsic_state

    def initialize(intrinsic_state)
        @intrinsic_state = intrinsic_state  # intrinsic state
    end

    def operation(extrinsic_state) # extrinsic state
        raise NotImplementedError, 'Subclasses must implement the operation method'
    end
end

# Concrete Flyweight class
class ConcreteFlyweight < Flyweight
    def operation(extrinsic_state)
        puts 'Operation inside concrete flyweight.'
    end
end

# FlyweightFactory class
class FlyweightFactory
    def initialize
        @flyweights = {}
    end

    def get_flyweight(intrinsic_state)
        @flyweights[intrinsic_state] ||= ConcreteFlyweight.new(intrinsic_state)
    end
end

# ClientClass
class ClientClass
    attr_reader :flyweight, :extrinsic_state

    def initialize(factory, intrinsic_state, extrinsic_state)
        @flyweight = factory.get_flyweight(intrinsic_state)
        @extrinsic_state = extrinsic_state
    end

    def operation
        puts 'Operation inside context.'
        flyweight.operation(extrinsic_state)
    end
end

# Client code
factory = FlyweightFactory.new
c = ClientClass.new(factory, 'common', 'separate1')
c.operation

c2 = ClientClass.new(factory, 'common', 'separate2')
c2.operation

puts c, c2
puts c.flyweight, c2.flyweight

=begin 
Operation inside context.
Operation inside concrete flyweight.
Operation inside context.
Operation inside concrete flyweight.
#<ClientClass:0x00007f307440e970>
#<ClientClass:0x00007f307440d2c8>
#<ConcreteFlyweight:0x00007f307440df98>
#<ConcreteFlyweight:0x00007f307440df98>
=end
